### PR TITLE
Normalization from example_ynh and some reworks

### PR DESCRIPTION
**Pull request by mastereur**
Packaging reworked to match the latest example_ynh standard.
Some improvements for the execution of installation and upgrade scripts.

### CODE REVIEW

**Comment by alexAubin** [comment]
Thank you for all the work ! :tada:

**Comment by kay0u** [comment]
:-O I was about to do it! Thank you!

File: hooks/post_user_create [hunk 1]
```diff
@@ -1,13 +1,21 @@
 #!/bin/bash
-app=YNH_APP
-src_path=/var/www/$app
-user=YNH_USER
-member=YNH_MEMBER
-
+app=dolibarr
```
Comment (kay0u) [inline hooks/post_user_create]:
```suggestion
app=__APP__
```

File: hooks/post_user_create [hunk 2]
```diff
@@ -1,13 +1,21 @@
...
+src_path=/var/www/dolibarr
```
Comment (kay0u) [inline hooks/post_user_create]:
```suggestion
src_path=__SRCPATH__
```

File: hooks/post_user_create [hunk 3]
```diff
@@ -1,13 +1,21 @@
...
+member=1
```
Comment (kay0u) [inline hooks/post_user_create]:
```suggestion
member=__MEMBER__
```

File: manifest.json [hunk 4]
```diff
@@ -1,49 +1,49 @@
...
         "install" : [
             {
                 "name": "domain",
-				"type": "domain",
+                "type": "domain",
                 "ask": {
-                    "en": "Choose a domain for Dolibarr",
-					"fr": "Choisissez un nom de domaine pour Dolibarr"
+                    "en": "Choose a domain name for ynhexample",
+                    "fr": "Choisissez un nom de domaine pour ynhexample"
```
Comment (kay0u) [inline manifest.json]:
```suggestion
                    "fr": "Choisissez un nom de domaine pour Dolibarr"
```

File: manifest.json [hunk 5]
```diff
@@ -1,49 +1,49 @@
...
             },
             {
                 "name": "path",
-				"type": "path",
+                "type": "path",
                 "ask": {
-                    "en": "Choose a path for Dolibarr",
-					"fr": "Choisissez un chemin pour Dolibarr"
+                    "en": "Choose a path for ynhexample",
+                    "fr": "Choisissez un chemin pour ynhexample"
```
Comment (kay0u) [inline manifest.json]:
```suggestion
                    "fr": "Choisissez un chemin pour Dolibarr"
```

File: scripts/_common.sh [hunk 6]
```diff
@@ -1,13 +1,20 @@
...
-	local checksum_setting_name=checksum_${1//[\/ ]/_}	# Replace all '/' and ' ' by '_'
-	ynh_app_setting_delete $app $checksum_setting_name
-}
\ No newline at end of file
+#=================================================
+# COMMON VARIABLES
+#=================================================
+
+# dependencies used by the app
+# pkg_dependencies="deb1 deb2"
```
Comment (kay0u) [inline scripts/_common.sh]: I think you can remove this line

File: scripts/install [hunk 7]
```diff
@@ -25,180 +16,341 @@ source /usr/share/yunohost/helpers
...
+### Use the execution time, given by --time, to estimate the weight of a step.
+### A common way to do it is to set a weight equal to the execution time in second +1.
+### The execution time is given for the duration since the previous call. So the weight should be applied to this previous call.
+ynh_script_progression --message="Validating installation parameters..." --time --weight=1
+
+### If the app uses nginx as web server (written in HTML/PHP in most cases), the final path should be "/var/www/$app".
+### If the app provides an internal web server (or uses another application server such as uwsgi), the final path should be "/opt/yunohost/$app"
 final_path=/var/www/$app
-test ! -e "$final_path" || ynh_die "This path already contains a folder"
+src_path=$final_path
```
Comment (kay0u) [inline scripts/install]: Why use `final_path` and `src_path` for the same thing? I noticed that you use either one or the other in others scripts.

I for me it's clearer to use only `final_path`, and use `${final_path}/htdocs/` where it's needed.

File: scripts/install [hunk 8]
```diff
@@ -25,180 +16,341 @@ source /usr/share/yunohost/helpers
...
+ynh_replace_string --match_string="__ADMIN__" --replace_string="$admin" --target_file="$finstall"
+if [ $member -eq 1 ]
+then
+	# If YNH users are members, we must activate the members module
+	ynh_replace_string --match_string="modLdap" --replace_string="modLdap,modAdherent" --target_file="$finstall"
+fi
+
+# Setup hooks
+fhook=../hooks/post_user_create
+ynh_replace_string --match_string="__ADMIN__" --replace_string="$admin" --target_file="$fhook"
```
Comment (kay0u) [inline scripts/install]: I don't see any \_\_ADMIN\_\_ in fhook

File: LICENSE [hunk 9]
```diff
@@ -1,23 +1,21 @@
-                     GNU GENERAL PUBLIC LICENSE
```
[the remark this answers is not in the document]
Comment (mastereur) [reply LICENSE]: I have the impression that this license is more appropriate for web applications with yunohost.
it says:"We recommend that people consider using the GNU AGPL for any software which will commonly be run over a network. "

_Auto translated from french to english with deepl_

File: scripts/install (hunk 7 above)
Comment (mastereur) [reply scripts/install]: I didn't understand the use of the final_path correctly.
From what I understand, it is used to define the folder where php files are placed.

File: scripts/_common.sh (hunk 6 above)
Comment (mastereur) [reply scripts/_common.sh]: Removed

File: hooks/post_user_create (hunk 1 above)
[the remark this answers is not in the document]
Comment (mastereur) [reply hooks/post_user_create]: C'est un résidu des tests. c'est corrigé

File: scripts/install (hunk 7 above)
Comment (kay0u) [reply scripts/install]: `final_path` is the path where your app is installed (what you call `src_path`).

You can change this line:
https://github.com/YunoHost-Apps/dolibarr_ynh/blob/7b42af1637a4e60c2954cc156bb51e3274fd3b00/conf/nginx.conf#L5

to: `alias __FINALPATH__/htdocs/`

To use this folder as the "php folder".

File: hooks/post_user_create (hunk 1 above)
[the remark this answers is not in the document]
Comment (kay0u) [reply hooks/post_user_create]: ok

File: LICENSE (hunk 9 above)
[the remark this answers is not in the document]
Comment (kay0u) [reply LICENSE]: I don't know if it's a good idea to change it now, but ok :-)

File: scripts/install (hunk 7 above)
Comment (mastereur) [reply scripts/install]: Great, I'll test this with all modifications by removing references to src_path

Comment (kay0u) [reply scripts/install]: Nice! If you need some help, do not hesitate!

File: README_fr.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
+# App exemple pour YunoHost
```
[the remark this answers is not in the document]
Comment (kay0u) [reply README_fr.md]: For me this one is not resolved.

File: scripts/install (hunk 7 above)
Comment (mastereur) [reply scripts/install]: j'ai fais les modifications avec les tests de non regression (install - backup - restore et upgrade) ça a l'air de bien fonctionner

Comment (kay0u) [reply scripts/install]: Everything looks good to me now!

Did you try your PR with [package_check] (https://github.com/YunoHost/package_check/) ?

If not, I can do it tonight!

Comment (mastereur) [reply scripts/install]: Unfortunately no, I didn't test with package_check, if you can do it?
thank you

File: scripts/upgrade [hunk 11]
```diff
@@ -8,88 +8,215 @@
...
-	sudo chown www-data: $lock
-
-	# Re-enable SSO
-	ynh_app_setting_delete $app unprotected_uris
-	sudo yunohost app ssowatconf
-
-	# Store the new version
-	ynh_app_setting_set $app version $version
+	cd $final_path/htdocs/install/
+	if php upgrade.php $current_version $update_version > /var/log/$app/upgrade.html;  then
```
Comment (kay0u) [inline scripts/upgrade]: Why a redirect into a .html here? (same for upgrade2, and upgrade3).

If it's normal, you should add

```bash
mkdir -p /var/log/$app/
```

 above this line.

File: check_process [hunk 12]
```diff
@@ -1,36 +1,36 @@
...
+		admin="john"	(USER)
+	; Checks
+		pkg_linter=1
+		setup_sub_dir=1
+		setup_root=1
+		setup_nourl=0
+		setup_private=0
+		setup_public=0
+		upgrade=1
+		upgrade=1	from_commit=CommitHash
```
Comment (kay0u) [inline check_process]:
```suggestion
		upgrade=1	from_commit=aced370f33a988d878be6434f1803ea68affe41a
```

The hash the current commit of master.

File: conf/install.forced.php [hunk 13]
```diff
@@ -14,31 +14,60 @@
...
+
+/** @var string Database server host */
 $force_install_dbserver = 'localhost';
+
+/** @var int Database server port */
 $force_install_port = 3306;
-$force_install_prefix = 'ynh_';
+
+/** @var string Database tables prefix */
+$force_install_prefix = 'llx_';
```
Comment (kay0u) [inline conf/install.forced.php]: Why did you change this line? Does this change couldn't break the upgrade from an old version?

File: check_process [hunk 14]
```diff
@@ -1,36 +1,36 @@
...
-    Level 9=0
-	Level 10=0
+	# If the level 5 (Package linter) is forced to 1. Please add justifications here.
+	Level 5=auto
+;;; Options
+Email=
+Notification=none
+;;; Upgrade options
+	; commit=CommitHash
+		name=Name and date of the commit.
```
Comment (kay0u) [inline check_process]:
```suggestion
		name=2 Feb 2018
```

File: scripts/change_url [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+	ynh_store_file_checksum --file="/etc/nginx/conf.d/$new_domain.d/$app.conf"
+fi
+
+#=================================================
+# SPECIFIC MODIFICATIONS
+#=================================================
+# UPDATE conf file
+#=================================================
+
+ynh_replace_string --match_string="$old_domain$old_path" --replace_string="$new_domain$new_path" --target_file="$final_path/conf/conf.php"
```
Comment (kay0u) [inline scripts/change_url]:
```suggestion
ynh_replace_string --match_string="$old_domain$old_path" --replace_string="$new_domain$new_path" --target_file="$final_path/htdocs/conf/conf.php"
```

File: scripts/upgrade [hunk 16]
```diff
@@ -8,88 +8,215 @@
...
+# GENERIC FINALIZATION
+#=================================================
+# SECURE FILES AND DIRECTORIES
+#=================================================
+
+# Set permissions on app files
+# Set permissions to app files
+chown -R root "$final_path"
+chmod 644 "$final_path/htdocs/conf/conf.php"
+chown -R $app "$datadir"
```
Comment (kay0u) [inline scripts/upgrade]: Make sure `$datadir` exists (it's missing during an upgrade from an old commit) :

```suggestion
mkdir -p $datadir
chown -R $app "$datadir"
```

File: scripts/upgrade (hunk 11 above)
Comment (mastereur) [reply scripts/upgrade]: ok, i will add this mkdir.
dolibarr documentation says : 

> If you want to script the database upgrade process (to avoid the upgrade process from web browser), you can run the 3 upgrade processes from command line: 
`$ cd htdocs/install $ php upgrade.php oldx.oldy.oldz x.y.z > output.html $ php upgrade2.php oldx.oldy.oldz x.y.z > output2.html $ php step5.php oldx.oldy.oldz x.y.z > output3.html`

File: conf/install.forced.php (hunk 13 above)
Comment (mastereur) [reply conf/install.forced.php]: llx_ is the default prefix of the application

Comment (kay0u) [reply conf/install.forced.php]: But the old value was 'ynh_' I see no reason to change it and I'm afraid that break upgrade. Anyway, I'll package check that!

Comment (mastereur) [reply conf/install.forced.php]: in principle, there should not be any problem during the upgrade because the prefix is set in the file htdocs/conf/conf.php

File: scripts/upgrade [hunk 17]
```diff
@@ -8,88 +8,218 @@
...
-	sudo chmod 444 $lock
-	sudo chown www-data: $lock
-
-	# Re-enable SSO
-	ynh_app_setting_delete $app unprotected_uris
-	sudo yunohost app ssowatconf
-
-	# Store the new version
-	ynh_app_setting_set $app version $version
+	cd $final_path/htdocs/install/
```
Comment (kay0u) [inline scripts/upgrade]:
```suggestion
	pushd $final_path/htdocs/install/
```

File: scripts/upgrade [hunk 18]
```diff
@@ -8,88 +8,218 @@
...
+		ynh_print_info --message="Step 2 upgrading ended successfully"
+	else
+		ynh_print_warn --message="Step 2 upgrading ended with error"
+	fi
+	if php step5.php $current_version $update_version > /var/log/$app/upgrade3.html; then
+		ynh_print_info --message="Step 3 upgrading ended successfully"
+	else
+		ynh_print_warn --message="Step 3 upgrading ended with error"
+	fi
+fi
```
Comment (kay0u) [inline scripts/upgrade]:
```suggestion
	popd
fi
```

This `push/popd` is to avoid some warning with helpers

**Comment by kay0u** [comment]
I'm sorry I always ask for something new on this PR. If I'm too boring, tell me and I'll make the changes myself.

**Comment by mastereur** [comment]
No, on the contrary, it's good to have a discussion. It gives you a good final code.

File: pull_request.md [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+
+## Solution
+- *easier yunohost integration*
+
+## PR Status
+- [ ] Code finished.
+- [ ] Tested with Package_check.
+- [ ] Fix or enhancement tested.
+- [ ] Upgrade from last version tested.
+- [  Can be reviewed and tested.
```
[the remark this answers is not in the document]
Comment (mastereur) [reply pull_request.md]: Corrected

File: scripts/upgrade [hunk 20]
```diff
@@ -8,88 +8,218 @@
...
+if [ "$upgrade_type" == "UPGRADE_APP" ]
+then
+	ynh_script_progression --message="Upgrading app from $current_version to $update_version" --time --weight=1
+
+	chown -R $app "$final_path"
 
 	# Remove the lock if it exists
-	lock=$src_path/documents/install.lock
+	lock=$final_path/documents/install.lock
 	[[ -f $lock ]] && sudo rm $lock
```
[the remark this answers is not in the document]
Comment (mastereur) [reply scripts/upgrade]: I modified according to your suggestions

[the remark this answers is not in the document]
Comment (kay0u) [reply scripts/upgrade]: Oups, sorry, I forget to surround the condition with `[` and `]` Which give:

```bash
if [ -f $lock ]
then
	ynh_secure_remove $lock
fi
```

[the remark this answers is not in the document]
Comment (mastereur) [reply scripts/upgrade]: Don't worry, I didn't see it either.

**Comment by mastereur** [comment]
With the new major version, I think we have to wait because there are always some mistakes.
We will do it for a version 10.0.X

**Comment by kay0u** [comment]
I will soon push an update normally ready to merge, I will drop the support of the old versions because it has been tagged as non-working and after some tests..... The old versions did not work:

https://github.com/YunoHost/apps/blob/84221aa48508026ee849cec2447c8663ac9118c2/apps.json#L396

**Comment by mastereur** [comment]
Great! Thank you for your work.
I didn't understand the importance of the weight of the messages, the page https://yunohost.org/#/packaging_apps_helpers is unavailable

**Comment by kay0u** [comment]
For me, it's ready to be merged, I just get a warning sometimes:
> configuring ldap
Warning: yunohost.hook <lambda> - [890.1] [WARN] ldap user update ended with error

But it's still working. If you have any idea how this problem can be avoided you are free to solve it...

Look at my code and tell me if we can merge it.

> I didn't understand the importance of the weight of the messages, the page https://yunohost.org/#/packaging_apps_helpers is unavailable

https://yunohost.org/#/packaging_apps_helpers_en  :-)

The weight define the proportion of the next action in the progress bar. The more the weight for an action is big, the more the progress bar will.... progress... at the end of this action (the end of an action is defined with the next call of `ynh_script_progression`).

The parameter `--time` is for the packager, with this one, the script will display the time taken by the previous action, so you can fill weights given value when you execute scripts.

**Comment by mastereur** [comment]
I just understood the error "ldap user update ended with error" when adding users, the script does not take into account the fact that the administrator is also a user and have his account already created.
This is not a problem because the administrator is created during installation (step 2)

Otherwise, yes, the code is clean and ready to be merged

**Comment by kay0u** [comment]
That was I thought too.

Before merging it, I would like to know if you want to be added to the maintainer of this repo?

You'll get permissions to commit/merge in this repo.

**Comment by mastereur** [comment]
yes why not. I will be a dolibarr user so having an updated version is one of the objectives. :)

**Comment by kay0u** [comment]
Thank you for this awesome work!

You just need to know that our convention is to have branches master with the last stable version and testing where you test new things like upgrade app...
